### ISSUE TITLE
What constitutes a "Omni" transaction

### ISSUE THREAD

**Issue by zathras-crypto** [OmniLayer/omnicore#183]
For feedback please, does a transaction with a Exodus/```omni``` marker but a non-decodable payload constitute an Omni transaction?

By non-decodable payload I mean that the transaction type and/or version are not known so we don't know how to decode the payload and ```interpret_Transaction``` returns false, leading to a ```PKT_ERROR -2``` from ```interpretPacket```.

For background, I'm trying to decide on how to address an issue I'm looking into - basically that Omni Core now records all transactions that get through initial parsing in ```txlistdb``` regardless of payload decoding, so let's say a transaction has an Exodus output/```omni``` marker but a non-decoable payload; this will be added to levelDB (```valid:false```).

Querying such a transaction then throws an error because the RPC code sees the transaction is in levelDB thus is an Omni transaction, but then fails to parse & decode it for the response.
```
omni_gettransaction cf0edfdd0a9e4d8aea2f87fc2cec6a367e1686a1fe066390ac195a797d76c0f1
Potential database corruption: Invalid transaction found (code -32603)
```

This is somewhat addressed in @dexX7's recent work for unconfirmed transactions, because the levelDB check is now skipped and thus when parsing & decoding fails the error thrown is:
```
Not a Master Protocol transaction
```

However having these non-decodable payloads added to levelDB as Omni transactions presents other side effects, such as ```omni_listblocktransactions``` which uses levelDB to look up all Omni transactions for a block.  It's not ideal behaviour to run ```omni_listblocktransactions```, get back a hash in the result showing there was a Omni transaction in the block, then query said hash & get a response that it's not a Omni transaction.

I hope this is making sense :)

TL:DR; are transactions with non-decodable payloads "Omni transactions" for the purposes of Omni Core.  If no, we should not add transactions with non-decodable payloads to levelDB.  If yes, querying the transaction should respond with some sort of base Omni tx object with unknown type/unknown version or similar rather than saying it's not an Omni transaction.

Any thoughts?

**Comment by dexX7** [comment]
> basically that Omni Core *now* records all transactions that get through initial parsing

Is this new behavior?

> It's not ideal behaviour to run omni_listblocktransactions, get back a hash in the result showing there was a Omni transaction in the block, then query said hash & get a response that it's not a Omni transaction.

In my opinion we should never ever provide the user with invalid transactions. I mentioned this also some time ago, when a similar question came up in the context of "list the number of transactions/trades" and whether it also count invalid transactions. I simply don't see any reason why.

In this particular case we have to differentiate though: Exodus purchases and DEx payments are Omni transactions, but we currently don't persist Exodus purchases, and use another processing path for DEx payments (i.e. no parse -> interpret, but instead DB lookups).

Short term: do we have to persist invalid transactions at all? If not, then let's get rid of them.

Longer term: classify and persist no-payload-Omni transactions properly. This is more tricky, because they don't really fit into the current model. I like the idea of having some wildcard entry ("unknown Omni transaction"), but it should not be used for invalid transactions imho.

**Comment by zathras-crypto** [comment]
> Is this new behavior?

0.0.9.3:
![screenshot from 2015-08-19 10 54 52](https://cloud.githubusercontent.com/assets/5633351/9346535/7a5c8bfa-4661-11e5-9a72-2b8a084ae583.png)

> In my opinion we should never ever provide the user with invalid transactions.

What exactly do you mean with invalid transactions? This is where it gets difficult, because language plays a big part hehe - do you mean:

* Invalid Omni transaction = bitcoin tx parsed as Omni with a decodable and understandable payload, but action rejected by state (eg invalid funds)
* Invalid Omni transaction = bitcoin tx parsed as Omni with a non-decodable payload

Because I think we should always tell users about invalid transactions of the first kind.  I kind of agree we shouldn't provide users with invalid transactions of the second kind.

> Short term: do we have to persist invalid transactions at all? If not, then let's get rid of them.

I'll do some further testing, but my gut feeling is we should be able to simply stop adding transactions with unknown type/versions to levelDB - it's a specific error code we get back for this scenario.  The point about Exodus & DEx is interesting, but I don't think it's an issue because by the time we're in packet interpretation we're already past Exodus/DEx processing.

**Comment by dexX7** [comment]
> What exactly do you mean with invalid transactions?

Ah, interesting, now I see where you're going.

So we basically have:

1. valid Omni transactions with payload, like simple sends
2. valid Omni transactions without payload, which we handle well, like DEx payments
3. valid Omni transactions without payload, which we don't handle well, like Exodus purchases
4. invalid Omni transactions, which are structually valid, but rejected, say due to insufficient funds
5. invalid Omni transactions, which are structurally invalid, like unknown transaction types, or transactions with only a marker, but nothing else
6. plain Bitcoin transactions

I need to refine my initial statement of "getting rid of *any* invalid transaction", because you indirectly pointed out something important, which I previously overlooked: there is a chance that users create transactions, which turn out to be invalid, whether that is, because they logged into another machine and resent some transactions, or due to our fault, because we didn't fully ensure a transaction (say via RPC) is also valid after confirmation, and there are probably more cases. This is relevant for the user, and should be visible.

Only slightly related, but I was a bit surprised: yesterday I went through the code and mentally tagged what parts are necessary for plain verification, to end up with the final state. The transaction, trade and STO DBs are not needed, and they only serve the RPC and UI layer.

This also means that we should probably primarily look at how we transfer and persist the data from the base layer, and make it available to the higher levels.

Maybe it helps (me to understand), if you could quickly clarify and outline the issue you currently see. It's not really clear to me how the following fits into the context exactly:

> ... my gut feeling is we should be able to simply stop adding transactions with unknown type/versions to levelDB [why do we care about those in particular?]. The point about Exodus & DEx is interesting, but I don't think it's an issue because by the time we're in packet interpretation we're already past Exodus/DEx processing [so?].

**Comment by dexX7** [comment]
Regarding the RPC queries:

We could also *not* throw a JSON error, which indicates that it's not an Omni transaction. It would then show the base fields like "txid", "valid":"false", ...

**Comment by zathras-crypto** [comment]
> Only slightly related, but I was a bit surprised: yesterday I went through the code and mentally tagged what parts are necessary for plain verification, to end up with the final state. The transaction, trade and STO DBs are not needed, and they only serve the RPC and UI layer.

Exactly - to be more specific they're only required to provide access to *history*.  If *now* (the current/final state) is all that is desired for a particular use-case and history can be discarded, then yep no DBs are needed at all.

> if you could quickly clarify and outline the issue you currently see.

The problem as I see it is that the following specific transaction scenario is currently added to levelDB.

> invalid Omni transactions, which are structurally invalid, like unknown transaction types, or transactions with only a marker, but nothing else

Which is in conflict with some other areas of the code, which assume that if a transaction is in levelDB that it is a structurally valid (but not necessary state-valid) transaction.  

I hope that's clear, but just shout if I'm still unclear :)

Quickly summarized, I think we should stop adding structurally invalid transactions to levelDB.

If payload decoding (```interpretTransaction```) fails ```PKT_ERROR - 2``` is returned, so a small patch like the following would do the job:

```diff
diff --git a/src/omnicore/omnicore.cpp b/src/omnicore/omnicore.cpp
index 62d21bd..2118532 100644
--- a/src/omnicore/omnicore.cpp
+++ b/src/omnicore/omnicore.cpp
@@ -2766,10 +2766,12 @@ bool mastercore_handler_tx(const CTransaction& tx, int nBlock, unsigned int idx,
         int interp_ret = mp_obj.interpretPacket();
         if (interp_ret) PrintToLog("!!! interpretPacket() returned %d !!!\n", interp_ret);
 
-        // only valid transactions get recorded
-        bool bValid = (0 <= interp_ret);
-        p_txlistdb->recordTX(tx.GetHash(), bValid, nBlock, mp_obj.getType(), mp_obj.getNewAmount());
-
+        // Only structurally valid transactions get recorded in levelDB
+        // PKT_ERROR - 2 = interpret_Transaction failed, structurally invalid payload
+        if (interp_ret != PKT_ERROR - 2) {
+            bool bValid = (0 <= interp_ret);
+            p_txlistdb->recordTX(tx.GetHash(), bValid, nBlock, mp_obj.getType(), mp_obj.getNewAmount());
+        }
         fFoundTx |= (interp_ret == 0);
     }
``` 

That change still passes all checkpoints, and seems to resolve the problem.  For example block 259439:
* contains zero structurally valid Omni transactions
* contains one structurally-invalid Omni transaction cf0edfdd0a9e4d8...

Before patch:
```
omni_listblocktransactions 259439
[
"cf0edfdd0a9e4d8aea2f87fc2cec6a367e1686a1fe066390ac195a797d76c0f1"
]
```

After patch:
```
omni_listblocktransactions 259439
[
]
```

**Comment by zathras-crypto** [comment]
Relevant PR https://github.com/OmniLayer/omnicore/pull/184

**Comment by dexX7** [comment]
Agreed, this is a pretty nice route. Does this resolve this issue?

Should we begin at some point to use a finer differentiation, when persisting data, and use more than one state (valid/invalid)?

**Comment by dexX7** [comment]
Since the initial issue was addressed, I'm going to close this, but feel free to reopen, if needed.
